### ISSUE TITLE
[Feature Request] Dropbox polling support 

### ISSUE THREAD

**Issue by ajkis** [rclone/rclone#2949]
It would be great if polling could be added for dropbox as well. 
API polling support: https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder-longpoll

**Comment by ajkis** [comment]
@jake-stripe is this something you could help with.
It would be great if dropbox users could have huge dir cache without manually refreshing every file/folder added to remote. I guess back then when dropbox was implemented API did not support polling, but now it does.

**Comment by ajkis** [comment]
@ncw any chance you add this one to v1.47 milestone.

**Comment by jake-stripe** [comment]
@ajkis I'm sorry but I'm pretty tied up with work right now. I won't be able work on this :/

**Comment by ajkis** [comment]
> @ajkis I'm sorry but I'm pretty tied up with work right now. I won't be able work on this :/

Thanks for reply.

**Comment by andresmatasuarez** [comment]
@jake-stripe @ajkis yesterday I raised [this issue](https://github.com/rclone/rclone/issues/4548) which is identical to this one. As I said, I can work on supporting poll-interval, I'd just need a little guidance as to where to start looking at in the code and to know which files are relevant to this.

**Comment by andresmatasuarez** [comment]
cc @ncw

**Comment by ncw** [comment]
@andresmatasuarez it would be great to have some help with this - sorry for not responding earlier!

If I was doing this, I'd take a look at the implementations of ChangeNotify we have in backend/drive and and add that into backend/dropbox.

https://github.com/rclone/rclone/blob/6a56ac1032bf03843850d1351d5239ef2960995b/backend/drive/drive.go#L2576-L2582

Here are the docs for the the dropbox API

https://www.dropbox.com/developers/documentation/http/documentation#files-list_folder-longpoll

It is certainly in the [dropbox SDK](https://godoc.org/github.com/dropbox/dropbox-sdk-go-unofficial/dropbox/files)

```go
    // ListFolderLongpoll : A longpoll endpoint to wait for changes on an
    // account. In conjunction with `listFolderContinue`, this call gives you a
    // low-latency way to monitor an account for file changes. The connection
    // will block until there are changes available or a timeout occurs. This
    // endpoint is useful mostly for client-side apps. If you're looking for
    // server-side notifications, check out our `webhooks documentation`
    // <https://www.dropbox.com/developers/reference/webhooks>.
    ListFolderLongpoll(arg *ListFolderLongpollArg) (res *ListFolderLongpollResult, err error)
```

That was super quick highspeed overview! Let me know if you want more.

**Comment by andresmatasuarez** [comment]
@ncw awesome, i'll try to kick this off sometime during this week, and I'll be letting you know of my progress as I make my way through the feature. As part of another personal project, I've already got to mess around with Dropbox API and this "list_folder#longpoll" method particularly, so I foresee I'll probably be much more occupied doing the integration into the `rclone` project, than experimenting with the API itself.

So, any extra, useful development/setup tip/advice you may consider worthwhile would be much appreciated :)

**Comment by ncw** [comment]
Great!

There is an overview of how to get set up here: https://rclone.org/install/#install-from-source

Setting up your editor to run `go fmt` on save of go files or better `goimports` is very useful. If you want a more fully featured IDE then set your editor up with [gopls](https://github.com/golang/tools/blob/master/gopls/README.md) - I use it with emacs and it is very useful if a bit memory intensive!

**Comment by jad3675** [comment]
I'm assuming this hasn't been implemented yet. I can't code myself out of a wet paper bag, but I'm more than happy to help with testing.

**Comment by andresmatasuarez** [comment]
@jad3675 I haven't started this yet but I'm organizing my time to get going

**Comment by andresmatasuarez** [comment]
@ncw I started to write this feature. I've finished setting up the whole go environment, but I'm having troubles with tests. 

I'm trying to run tests for `backend/drive` just to check everything was set up correctly but I keep getting the following when running `sudo go test -v ./backend/drive` (from the root of the project):
```
...
=== RUN   TestIntegration
    fstests.go:396: Using remote "TestDrive:"
    fstests.go:417: Didn't find "TestDrive:" in config file - skipping tests
...
```

I'm prepending `sudo` to the test command because otherwise I'll get:
```
=== RUN   TestIntegration
2020/12/21 03:39:50 Failed to load config file "/home/amatasuarez/.config/rclone/rclone.conf": open /home/amatasuarez/.config/rclone/rclone.conf: permission denied
FAIL    github.com/rclone/rclone/backend/drive  0.009s
FAIL
```

I sort of did my homework and gave a thorough look at https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#testing and created a new remote `TestDrive` (I connected an old gmail account of mine with no files in it, completely disposable). I did that via `sudo rclone config create TestDrive drive` with auto config enabled.

But tests still don't seem to find this new `TestDrive` remote because they keep failing with the above error.

I'm a pretty newbie at go so maybe this is too obvious, but I'd really appreciate some help here so I can get down to the code right away

**Comment by jad3675** [comment]
> @ncw I started to write this feature. I've finished setting up the whole go environment, but I'm having troubles with tests.
> 
> I'm trying to run tests for `backend/drive` just to check everything was set up correctly but I keep getting the following when running `sudo go test -v ./backend/drive` (from the root of the project):
> 
> ```
> ...
> === RUN   TestIntegration
>     fstests.go:396: Using remote "TestDrive:"
>     fstests.go:417: Didn't find "TestDrive:" in config file - skipping tests
> ...
> ```
> 
> I'm prepending `sudo` to the test command because otherwise I'll get:
> 
> ```
> === RUN   TestIntegration
> 2020/12/21 03:39:50 Failed to load config file "/home/amatasuarez/.config/rclone/rclone.conf": open /home/amatasuarez/.config/rclone/rclone.conf: permission denied
> FAIL    github.com/rclone/rclone/backend/drive  0.009s
> FAIL
> ```
> 
> I sort of did my homework and gave a thorough look at https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#testing and created a new remote `TestDrive` (I connected an old gmail account of mine with no files in it, completely disposable). I did that via `sudo rclone config create TestDrive drive` with auto config enabled.
> 
> But tests still don't seem to find this new `TestDrive` remote because they keep failing with the above error.
> 
> I'm a pretty newbie at go so maybe this is too obvious, but I'd really appreciate some help here so I can get down to the code right away

Use the --config flag on the rclone command to specify the config file that has 'TestDrive' remote in it.

John

**Comment by andresmatasuarez** [comment]
I'm doing `go test -v ./backend/drive`, where should I place the `--config` param? `go test` doesn't seem to take in anything like that

**Comment by jad3675** [comment]
> @jad3675 I'm doing `go test -v ./backend/drive`, where should I place the `--config` param? `go test` doesn't seem to take in anything like that

Hm. Nevermind, I mis-read your original post and/or it was too early in the morning.
Assuming you're running go test as yourself,  go test should have permission to /home/amatasuarez/.config/rclone/rclone.conf. What are the perms on that file? 
Since you created it with 'sudo rclone config create TestDrive drive' the user permissions are might be wrong. You can chown it with your own user and group

John

**Comment by andresmatasuarez** [comment]
@jad3675  thanks for taking the time

I'm also running `go test` with `sudo`, shouldn't it be able to easily read `~/.config/rclone/rclone.conf`?

**Comment by jad3675** [comment]
> @jad3675 thanks for taking the time
> 
> I'm also running `go test` with `sudo`, shouldn't it be able to easily read `~/.config/rclone/rclone.conf`?

It's trying to read (I assume) /root/.config/rclone/rclone.conf, which isn't there. You shouldn't need to run go test as sudo - it has to be the permissions on the rclone.conf file since it was updated with sudo. Changing the user/group on ~/.config/rclone/rclone.conf to what it should be should fix it right up.

**Comment by andresmatasuarez** [comment]
@jad3675 you were right, that absolutely did it, huge thanks

I've been writing down code since then and reached a point where I have every added/updated/deleted file entry.

I've been using Amazon CloudDrive and Drive `ChangeNotify` implementations as reference and now I've a couple of interrelated questions.

1) What's the rough idea behind what I need to do? I'm not asking for a specific, line-by-line explanation, just the overall pseudo-algorithm, like "now that you have all the file change entries, you should then do this and that and make sure that this ends up like that. Once this is done, you'd have to call `notifyFunc` with X and Y". This would make the reference code easier to follow and make me write better stuff.

2) By looking at both `drive.go`'s and `amazonclouddrive.go`'s implementations, I see this:
![image](https://user-images.githubusercontent.com/6817802/103029525-bf9aa400-4538-11eb-8e83-5e80316274bd.png)

If I had to guess, what that's doing is iterating over all the changes polled in and sort of invalidating those paths in the cache. To replicate the same in `dropbox.go` I'd have to declare and assign `f.dirCache` as it's done in one of the above files. Thing is the code initializing this value is different in both cases and I believe depends on my knowing a broader portion of `rclone` codebase. I'd be delighted if you could provide me with a rough idea of how I need to instantiate `f.dirCache` and a brief description of what it does, same as question 1).

thanks in advance

**Comment by andresmatasuarez** [comment]
@john

**Comment by ncw** [comment]
> 1. What's the rough idea behind what I need to do? I'm not asking for a specific, line-by-line explanation, just the overall pseudo-algorithm, like "now that you have all the file change entries, you should then do this and that and make sure that this ends up like that. Once this is done, you'd have to call `notifyFunc` with X and Y". This would make the reference code easier to follow and make me write better stuff.

What you have to do is send a notify message for each file or directory that changes. A lot of the complication in the drive/acd backends are to do with turning the ID of the object into a file path which is what rclone needs in the notify message.

In the messages returned from longpoll/continue you get this

            "path_display": "/Homework/math/Prime_Numbers.txt",

which is the path that rclone needs. You need to chop the "root" off it - if you remove the `f.slashRootSlash` from the start of the path_display that should be the path rclone needs. It can be empty - that just means the root.

> If I had to guess, what that's doing is iterating over all the changes polled in and sort of invalidating those paths in the cache.

That is it.

> To replicate the same in `dropbox.go` I'd have to declare and assign `f.dirCache` as it's done in one of the above files. Thing is the code initializing this value is different between files and I believe depends on my knowing a broader portion of `rclone` codebase. I'd be delighted if you could provide me with a rough idea of how to instantiate `f.dirCache` and a brief description of what it does, same as question 1).

You don't need a dirCache, you just need to use the display_path with the root chopped off in the message.

Hopefully that makes sense!

**Comment by wolveix** [comment]
@andresmatasuarez have you had the chance to make any additional progress on this implementation? 😃

**Comment by ncw** [comment]
Thanks to excellent work by @wolveix this is now implemented :-)

This is now merged into master which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.55

Testing much appreciated :-)

**Comment by andresmatasuarez** [comment]
@ivandeex @ncw @wolveix amazing work! I'd like to apologize to all of you for relying on me to do this, but I hadn't had the time to resume extracurricular projects until now :disappointed: 

Anyway, I just installed the latest beta to see dropbox polling in action and I'm getting a couple of potentials errors related to it which I'm explaining below:

I've given rclone the Dropbox's "app folder" type access, meaning I've scoped access to a certain sub-folder in my cloud drive, let's assume it's `Apps/my-media`

I'm mounting my cloud files in the following way:
`rclone mount {...}  dropbox-visions-of-perversity:/ /media-from-dropbox/`

Note the `/` beside the colon since I want everything that's inside `Apps/my-media` to be mounted.

However, I'm getting the following in the logs:
`INFO  : Dropbox root '': Failed to get StartCursor: Error in call to API function "files/list_folder/get_latest_cursor": Path root is not supported for sandbox app.`

If I change my command to `rclone mount {...}  dropbox-visions-of-perversity: /media-from-dropbox/` (removed `/` next to colon), I get instead:
```
INFO  : Dropbox root '': Failed to get StartCursor: Error in call to API function "files/list_folder/get_latest_cursor": request body: path: Specify the root folder as an empty string rather than as "/".
```

...and we're back to square one.

I found some insight in this support ticket: https://www.dropboxforum.com/t5/Dropbox-API-Support-Feedback/Download-on-scoped-app/td-p/462689

If I had to guess, I'd say the problem lies [in this line](https://github.com/rclone/rclone/pull/5049/files#diff-7070c850317ae70144b7c0544c568a183ab48d05f0210e96160c3ecf3746ff22R1258), particularly in the portion:
`f.opt.Enc.FromStandardPath(f.slashRoot)`
